test/test_vmap.py

```diff
@@ -999,6 +995,66 @@ def foo(x):
                 )""")
         self.assertEqual(buf, expected)
 
+    def _test_vmap_autocast(self, device):
+
+        if torch.device(device).type == "cpu":
+            amp_dtype = torch.bfloat16
+        else:
+            amp_dtype = torch.float16
+
+        a_float32 = torch.rand(4, 2, 3, device=device)
+        b_float32 = torch.rand(4, 3, 2, device=device)
+        c_float32 = torch.rand(4, 2, 2, device=device)
+        d_float32 = torch.rand(4, 3, 2, device=device)
+
+        # Case 1, autocast inside vmapped function
+        def func1(x, y, z, w):
+            with torch.autocast(dtype=amp_dtype, device_type=device):
+                e_float16 = torch.matmul(x, y)
+                assert e_float16.dtype == amp_dtype, e_float16.dtype
+                f_float16 = torch.matmul(z, e_float16)
+                assert f_float16.dtype == amp_dtype, f_float16.dtype
+            return torch.matmul(w, f_float16.float())
+
+        expected = func1(a_float32, b_float32, c_float32, d_float32)
+        out = vmap(func1)(a_float32, b_float32, c_float32, d_float32)
+        assert expected.allclose(out)
+
+        # Case 2, autocast decorator inside vmapped function
+        @torch.autocast(dtype=amp_dtype, device_type=device)
+        def func2(x, y, z, w):
+            e_float16 = torch.matmul(x, y)
+            assert e_float16.dtype == amp_dtype, e_float16.dtype
+            f_float16 = torch.matmul(z, e_float16)
+            assert f_float16.dtype == amp_dtype, f_float16.dtype
+            return torch.matmul(w, f_float16)
+
+        expected = func2(a_float32, b_float32, c_float32, d_float32)
+        out = vmap(func2)(a_float32, b_float32, c_float32, d_float32)
+        assert expected.allclose(out)
+
+        # Case 3, autocast is outside vmapped function
+        def func3(x, y, z, w):
+            e_float16 = torch.matmul(x, y)
+            assert e_float16.dtype == amp_dtype, e_float16.dtype
+            f_float16 = torch.matmul(z, e_float16)
+            assert f_float16.dtype == amp_dtype, f_float16.dtype
+            return torch.matmul(w, f_float16)
+
+        with torch.autocast(dtype=amp_dtype, device_type=device):
+            expected = func3(a_float32, b_float32, c_float32, d_float32)
+            out = vmap(func3)(a_float32, b_float32, c_float32, d_float32)
+
+        assert expected.allclose(out)
+
+    @skip("Somehow, vmap and autocast do not work on CPU")
+    def test_vmap_autocast_cpu(self):
+        self._test_vmap_autocast("cpu")
```

Comment:
Interesting. Do you have a stack trace?

Reply:
The reason why this test is failing is that matmul uses decomposes to bmm which has at::matmul in the batching rule.
But matmul with autocast on CPU only goes to fallthrough: https://github.com/pytorch/pytorch/blob/c76c6e9bd345ba697ea2a6fb8c2e24f051bb2e00/aten/src/ATen/autocast_mode.cpp#L465-L480 and autocasted on GPU : https://github.com/pytorch/pytorch/blob/c76c6e9bd345ba697ea2a6fb8c2e24f051bb2e00/aten/src/ATen/autocast_mode.cpp#L331